scapy/arch/pcapdnet.py

```diff
@@ -372,38 +360,58 @@ def send(self, x):
     conf.L2socket = L2pcapSocket
     conf.L3socket = L3pcapSocket
 
+################
+#  PCAP/PCAPY  #
+################
+
+
 if conf.use_pcap:
     try:
-        import pcap
+        import pcap  # python-pypcap
+        _PCAP_MODE = "pypcap"
     except ImportError as e:
         try:
-            import pcapy as pcap
+            import libpcap as pcap  # python-libpcap
+            _PCAP_MODE = "libpcap"
         except ImportError as e2:
-            if conf.interactive:
-                log_loading.error("Unable to import pcap module: %s/%s", e, e2)
-                conf.use_pcap = False
-            else:
-                raise
+            try:
+                import pcapy as pcap  # python-pcapy
+                _PCAP_MODE = "pcapy"
+            except ImportError as e3:
+                if conf.interactive:
+                    log_loading.error("Unable to import pcap module: %s/%s", e, e2)
+                    conf.use_pcap = False
+                else:
+                    raise
     if conf.use_pcap:
 
         # From BSD net/bpf.h
         # BIOCIMMEDIATE=0x80044270
         BIOCIMMEDIATE = -2147204496
 
-        if hasattr(pcap, "pcap"):  # python-pypcap
+        if _PCAP_MODE == "pypcap":  # python-pypcap
             class _PcapWrapper_pypcap:
-                def __init__(self, device, snaplen, promisc, to_ms):
+                def __init__(self, device, snaplen, promisc, to_ms, monitor=False):
                     try:
-                        self.pcap = pcap.pcap(device, snaplen, promisc, immediate=1, timeout_ms=to_ms)
+                        self.pcap = pcap.pcap(device, snaplen, promisc, immediate=1, timeout_ms=to_ms, rfmon=monitor)
                     except TypeError:
-                        # Older pypcap versions do not support the timeout_ms argument
-                        self.pcap = pcap.pcap(device, snaplen, promisc, immediate=1)
+                        try:
+                            if monitor:
+                                warning("Your pypcap version is too old to support monitor mode, Please update !")
+                            self.pcap = pcap.pcap(device, snaplen, promisc, immediate=1, timeout_ms=to_ms)
+                        except TypeError:
+                            # Even older pypcap versions do not support the timeout_ms argument
+                            self.pcap = pcap.pcap(device, snaplen, promisc, immediate=1)
 
                 def __getattr__(self, attr):
                     return getattr(self.pcap, attr)
 
                 def __del__(self):
-                    warning("__del__: don't know how to close the file descriptor. Bugs ahead ! Please report this bug.")
+                    try:
+                        self.pcap.close()
+                    except AttributeError:
+                        warning("__del__: don't know how to close the file "
+                                "descriptor. Bugs ahead! Please update python-pypcap!")
```

Comment:
Same here, could you specify the version?

Reply:
Done (they published 1.2.1)